devices/src/virtio/vsock/csm/connection.rs

```diff
@@ -0,0 +1,632 @@
+    fn get_polled_fd(&self) -> RawFd {
+        self.stream.as_raw_fd()
+    }
+
+    /// Get the event set that this connection is interested in.
+    ///
+    /// A connection will want to be notified when:
+    /// - data is available to be read from the host stream, so that it can store an RW pending
+    ///   RX indication; and
+    /// - data can be written to the host stream, and the TX buffer needs to be flushed.
+    ///
+    fn get_polled_evset(&self) -> epoll::Events {
+        let mut evset = epoll::Events::empty();
+        if !self.tx_buf.is_empty() {
+            // There's data waiting in the TX buffer, so we are interested in being notified when
+            // writing to the host stream wouldn't block.
+            evset.insert(epoll::Events::EPOLLOUT);
+        }
+        // We're generally interested in being notified when data can be read from the host stream,
+        // unless we're in a state which doesn't allow moving data from host to guest.
+        match self.state {
+            ConnState::LocalClosed | ConnState::PeerClosed(true, _) => (),
+            _ if self.need_credit_update_from_peer() => (),
+            _ => evset.insert(epoll::Events::EPOLLIN),
+        }
+        evset
+    }
+
+    /// Notify the connection about an event (or set of events) that it was interested in.
+    ///
+    fn notify(&mut self, evset: epoll::Events) {
+        if evset.contains(epoll::Events::EPOLLIN) {
+            // Data can be read from the host stream. Setting a Rw pending indication, so that the
+            // muxer will know to call `recv_pkt()` later.
+            self.pending_rx.insert(PendingRx::Rw);
+        }
+
+        if evset.contains(epoll::Events::EPOLLOUT) {
+            // Data can be written to the host stream. Time to flush out the TX buffer.
+            //
+            if self.tx_buf.is_empty() {
+                info!("vsock: connection received unexpected EPOLLOUT event");
+                return;
+            }
+            let flushed = self
+                .tx_buf
+                .flush_to(&mut self.stream)
+                .unwrap_or_else(|err| {
+                    info!(
+                        "vsock: error flushing TX buf for (lp={}, pp={}): {:?}",
+                        self.local_port, self.peer_port, err
+                    );
+                    self.kill();
+                    0
+                });
+            self.fwd_cnt += Wrapping(flushed as u32);
+
+            // If we've freed up some more buffer space, we may need to let the peer know it can
+            // safely send more data our way.
+            if self.peer_needs_credit_update() {
+                self.pending_rx.insert(PendingRx::CreditUpdate);
+            }
+
+            // If this connection was shutting down, but is wating to drain the TX buffer
+            // before forcefull termination, the wait might be over.
+            if self.state == ConnState::PeerClosed(true, true) && self.tx_buf.is_empty() {
```

Comment:
Is it possible to have the `will_not_send_anymore_data` flag on and be wating to drain the TX buffer `tx_buf`?